### PR TITLE
Refactor: RandomActivity to use ViewModel

### PR DESCRIPTION
**Pull request by Lampotrias**
Hi all!

It's my version refactoring RandomActivity with ViewModel.
https://phabricator.wikimedia.org/T303942

Thanks in advance for the code review!

### CODE REVIEW

File: app/src/main/java/org/wikipedia/random/RandomViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+class RandomViewModel(bundle: Bundle) : ViewModel() {
+
+	private val funnel: RandomizerFunnel
+	private var saveShareJob: Job? = null
+
+	private val handler = CoroutineExceptionHandler { _, throwable ->
+		L.w(throwable)
+	}
+
+	private val _savaShareState = MutableStateFlow(false)
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/random/RandomViewModel.kt:27]: misspelling

File: app/src/main/java/org/wikipedia/random/RandomViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	val savaShareState: StateFlow<Boolean> = _savaShareState
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/random/RandomViewModel.kt:28]: misspelling

File: app/src/main/java/org/wikipedia/random/RandomViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+	private val _saveToDefaultList = MutableStateFlow<SaveSharedState<PageTitle>>(Initial())
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/random/RandomViewModel.kt:30]: What's the reason for having underscore `_` and non-underscore versions of all these fields?

File: app/src/main/java/org/wikipedia/random/RandomViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		_saveToDefaultList.value = Result(title)
+	}
+
+	fun saveToCustomList(title: PageTitle) {
+		_saveToCustomList.value = Result(title)
+	}
+
+	fun movePageToList(sourceReadingListId: Long, title: PageTitle) {
+		_movePageToList.value = Result(MoveDataSource(sourceReadingListId, title))
+	}
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/random/RandomViewModel.kt:56]: I'm not sure I see the benefit of the above three functions. You're passing in a PageTitle, which gets passed to a Flow, which gets passed right back to the Fragment. Why not keep these interactions within the Fragment as before?

File: app/src/main/java/org/wikipedia/random/RandomViewModel.kt (hunk 3 above)
Comment (Lampotrias) [reply app/src/main/java/org/wikipedia/random/RandomViewModel.kt:30]: The non-underscored fields used only observe values in other entities, such fragments or activities. The underscored field guarantees emit values only from view model. Unfortunately now not other way for it.

File: app/src/main/java/org/wikipedia/random/RandomViewModel.kt (hunk 4 above)
Comment (Lampotrias) [reply app/src/main/java/org/wikipedia/random/RandomViewModel.kt:56]: I'll fix it. I created these methods with the expectation that the logic could be complicated in the future

**Comment by Lampotrias** [comment]
Thank you for your feedback.

I fixed all defects in pull request. In RandomItemFragment was added viewmodel and removed `retainInstance = true`. I tested case with rotation of screen and saw no problem with it.

**Review by dbrant — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/random/RandomItemViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import org.wikipedia.dataclient.WikiSite
+import org.wikipedia.dataclient.page.PageSummary
+import org.wikipedia.util.Resource
+
+class RandomItemViewModel(bundle: Bundle) : ViewModel() {
+
+	private val wikiSite: WikiSite = bundle.getParcelable(RandomActivity.INTENT_EXTRA_WIKISITE)!!
+
+	private val _saveShareState = MutableStateFlow<SaveSharedState>(Initial)
+	val saveShareState = _saveShareState.asStateFlow()
```
[the remark this answers is not in the document]
Comment (Lampotrias) [reply app/src/main/java/org/wikipedia/random/RandomItemViewModel.kt:21]: Yes, you are right.

File: app/src/main/java/org/wikipedia/random/RandomItemFragment.kt [hunk 6]
```diff
@@ -33,7 +34,7 @@ class RandomItemFragment : Fragment() {
     private var _binding: FragmentRandomItemBinding? = null
     private val binding get() = _binding!!
 
-    private val disposables = CompositeDisposable()
+    private val viewModel: RandomItemViewModel by viewModels { RandomItemViewModel.Factory(requireArguments()) }
 
     private lateinit var wikiSite: WikiSite
```
[the remark this answers is not in the document]
Comment (Lampotrias) [reply app/src/main/java/org/wikipedia/random/RandomItemFragment.kt:39]: Now, I don't need `wikiSite`  in the ViewModel, I planning use `WikipediaApp.getInstance().wikiSite`. Currently, `wikiSite` use in `RandomItemFragment` for call:
```L10nUtil.setConditionalLayoutDirection(view, wikiSite.languageCode)```, as it was before the refactoring

File: app/src/main/java/org/wikipedia/random/RandomItemViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+	private val disposables = CompositeDisposable()
+
+	init {
+		getRandomPage()
+	}
+
+	fun getRandomPage() {
+		val d = ServiceFactory.getRest(wikiSite).randomSummary
```
[the remark this answers is not in the document]
Comment (Lampotrias) [reply app/src/main/java/org/wikipedia/random/RandomItemViewModel.kt:30]: I have question with select MediaWiki Action API. Do I need to make a request like `https://en.wikipedia.org/w/api.php?action=query&format=json&list=random&formatversion=2` and then get an instance of PageSummary

File: app/src/main/java/org/wikipedia/random/RandomItemFragment.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (Lampotrias) [reply app/src/main/java/org/wikipedia/random/RandomItemFragment.kt]: Ok, now `wikiSite` is given from `viewModel`.

**Comment by Lampotrias** [comment]
Thank you for your feedback.

I removed all code, that was created for the future, such Factory.

And sorry about the copy-paste code. This is my first PR, I'm a little worried.

**Pull request closed without merging**